futurewei-cloud/arktos
# k8s endpoints read-only sharing: able to get, unable to create or update

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR changes Endpoints resource get code, when getting "default" ns "kubernetes-xxx" endpoints, returns the content based on the system default/kubernetes object.

Based on [Endpoints section, multi-tenancy networking design proposal](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/multi-tenancy/multi-tenancy-network.md#endpoints), _Query for default-ns scoped kubernetes-{network} shall get back the proper content based on the cluster kubernetes endpoints object. System does not duplicate such endpoints; instead it derives content based on the cluster kubernetes endpoints._

As these endpoints are like read-only views of the system kubernetes EP, any write/delete through them is disallowed.

This PR covers CRU (Delete does not need code change). This PR does not cover LW. LW support will be considered at client side at informer code instead, at current stage.

**How to verify at local dev env**:
1. start etcd & kube-apiserver
2. verify the system kubernetes EP is able to get
```
kubectl get ep kubernetes
NAME         ENDPOINTS           AGE   SERVICEGROUPID
kubernetes   192.168.0.18:6443   13d   0
```
3. verify able to get its read-only EP in another tenant
```
kubectl create tenant x
kubectl get ep --tenant x kubernetes-foo
NAME             ENDPOINTS           AGE   SERVICEGROUPID
kubernetes-foo   192.168.0.18:6443   13d   0
```
4. verify unable to create such read-only EP, by running ```kubectl create -f k8s-ep.yaml```, with k8s-ep.yaml has content
```yaml
apiVersion: v1
kind: Endpoints
metadata:
  name: kubernetes-bar
  namespace: default
  tenant: x
  labels:
    arktos.futurewei.com/network: bar
subsets:
- addresses:
  - ip: 192.168.0.18
  ports:
  - name: https
    port: 6443
    protocol: TCP
```
the error returned is ```The Endpoints "kubernetes-bar" is invalid: kubernetes-bar: Forbidden: read only resource not allowed to create or update.```

**Does this PR introduce a user-facing change?**:
Yes
```release-note
Endpoints in default namespace with name pattern starting with "kubernetes-" is considered as candidates of kubernetes read-only endpoints of network (kubernetes-alias EP in short). They are not allowed to create/update/delete, the returned content to get is based on the system default/kubernetes EP. 
```
